File: ambeth/Ambeth.Cache/ambeth/proxy/CachePostProcessor.cs
using System;
using System.Collections.Generic;
using De.Osthus.Ambeth.Annotation;
using De.Osthus.Ambeth.Cache.Interceptor;
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Config;
using De.Osthus.Ambeth.Ioc.Factory;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Service;

namespace De.Osthus.Ambeth.Proxy
{
    public class CachePostProcessor : AbstractCascadePostProcessor
    {
        [LogInstance]
        public ILogger Log { private get; set; }

        public class ServiceAnnotationCache : AnnotationCache<ServiceAttribute>
        {
            protected override bool AnnotationEquals(ServiceAttribute left, ServiceAttribute right)
            {
                return Equals(left.Interface, right.Interface) && Equals(left.Name, right.Name);
            }
        }

        public class ServiceClientAnnotationCache : AnnotationCache<ServiceClientAttribute>
        {
            protected override bool AnnotationEquals(ServiceClientAttribute left, ServiceClientAttribute right)
            {
                return Equals(left.Name, right.Name);
            }
        }

        protected AnnotationCache<ServiceAttribute> serviceAnnotationCache = new ServiceAnnotationCache();

        protected AnnotationCache<ServiceClientAttribute> serviceClientAnnotationCache = new ServiceClientAnnotationCache();

        [Property(ServiceConfigurationConstants.NetworkClientMode, DefaultValue = "false")]
        public bool IsNetworkClientMode { protected get; set; }

        protected override ICascadedInterceptor HandleServiceIntern(IBeanContextFactory beanContextFactory, IServiceContext beanContext, IBeanConfiguration beanConfiguration, Type type, ISet<Type> requestedTypes)
        {
            ServiceAttribute serviceAnnotation = serviceAnnotationCache.GetAnnotation(type);
            if (serviceAnnotation != null)
            {
                return HandleServiceAnnotation(serviceAnnotation, beanContextFactory, beanContext, beanConfiguration, type);
            }
            AnnotationEntry<ServiceClientAttribute> serviceClientAnnotation = serviceClientAnnotationCache.GetAnnotationEntry(type);
            if (serviceClientAnnotation != null)
            {
                return HandleServiceClientAnnotation(serviceClientAnnotation, beanContextFactory, beanContext, beanConfiguration, type);
            }
            return null;
        }

        protected String ExtractServiceName(String serviceName, Type type)
        {
            if (serviceName == null || serviceName.Length == 0)
            {
                serviceName = type.Name;
                if (serviceName.EndsWith("Proxy"))
                {
                    serviceName = serviceName.Substring(0, serviceName.Length - 5);
                }
                if (serviceName[0] == 'I' && Char.IsUpper(serviceName[1]))
                {
                    serviceName = serviceName.Substring(1);
                }
            }
            return serviceName;
        }

        protected ICascadedInterceptor HandleServiceAnnotation(ServiceAttribute serviceAnnotation, IBeanContextFactory beanContextFactory, IServiceContext beanContext, IBeanConfiguration beanConfiguration, Type type)
        {
            if (serviceAnnotation.CustomExport)
            {
                // Do nothing if the service wants to be exported by some special way anywhere else
                return null;
            }
            String beanName = beanConfiguration.GetName();
            String serviceName = ExtractServiceName(serviceAnnotation.Name, type);
            if (!IsNetworkClientMode)
            {
                CacheInterceptor interceptor = new CacheInterceptor();
                if (beanContext.IsRunning)
                {
                    interceptor = beanContext.RegisterWithLifecycle(interceptor).PropertyValue("ServiceName", serviceName).IgnoreProperties("ProcessService").Finish();
                    beanContext.Link(beanName).To<IServiceExtendable>().With(serviceName);
                }
                else
                {
                    beanContextFactory.RegisterWithLifecycle(interceptor).PropertyValue("ServiceName", serviceName).IgnoreProperties("ProcessService");
                    beanContextFactory.Link(beanName).To<IServiceExtendable>().With(serviceName);
                }
                if (Log.InfoEnabled)
                {
                    Log.Info("Registering application service '" + serviceName + "'");
                }
                return interceptor;
            }
            else
            {
                if (Log.InfoEnabled)
                {
                    Log.Info("Registering application mock service '" + serviceName + "'");
                }
                if (beanContext.IsRunning)
                {
                    beanContext.Link(beanName).To<IServiceExtendable>().With(serviceName);
                }
                else
                {
                    beanContextFactory.Link(beanName).To<IServiceExtendable>().With(serviceName);
                }
                return null;
            }
        }

        protected ICascadedInterceptor HandleServiceClientAnnotation(AnnotationEntry<ServiceClientAttribute> serviceClientAnnotation, IBeanContextFactory beanContextFactory, IServiceContext beanContext,
            IBeanConfiguration beanConfiguration, Type type)
        {
            String serviceName = ExtractServiceName(serviceClientAnnotation.Annotation.Name, serviceClientAnnotation.DeclaringType);
            CacheInterceptor interceptor = new CacheInterceptor();
            if (beanContext.IsRunning)
            {
                interceptor = beanContext.RegisterWithLifecycle(interceptor).PropertyValue("ServiceName", serviceName).Finish();
            }
            else
            {
                beanContextFactory.RegisterWithLifecycle(interceptor).PropertyValue("ServiceName", serviceName);
            }

            if (Log.InfoEnabled)
            {
                Log.Info("Creating application service stub for service '" + serviceName + "' accessing with '" + serviceClientAnnotation.DeclaringType.FullName + "'");
            }
            return interceptor;
        }

        protected String BuildCacheInterceptorName(String serviceName)
        {
            return "cacheInterceptor." + serviceName;
        }
    }
}

File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/proxy/CachePostProcessor.java
package de.osthus.ambeth.proxy;

import java.lang.annotation.Annotation;
import java.lang.reflect.AnnotatedElement;
import java.util.Set;

import de.osthus.ambeth.annotation.AnnotationCache;
import de.osthus.ambeth.annotation.AnnotationEntry;
import de.osthus.ambeth.cache.Cached;
import de.osthus.ambeth.cache.interceptor.CacheInterceptor;
import de.osthus.ambeth.config.Property;
import de.osthus.ambeth.config.ServiceConfigurationConstants;
import de.osthus.ambeth.ioc.IServiceContext;
import de.osthus.ambeth.ioc.config.IBeanConfiguration;
import de.osthus.ambeth.ioc.factory.IBeanContextFactory;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.service.IServiceExtendable;
import de.osthus.ambeth.util.EqualsUtil;

public class CachePostProcessor extends MergePostProcessor
{
	@LogInstance
	private ILogger log;

	protected final AnnotationCache<Service> serviceAnnotationCache = new AnnotationCache<Service>(Service.class)
	{
		@Override
		protected boolean annotationEquals(Service left, Service right)
		{
			return EqualsUtil.equals(left.value(), right.value()) && EqualsUtil.equals(left.name(), right.name());
		}
	};

	protected final AnnotationCache<ServiceClient> serviceClientAnnotationCache = new AnnotationCache<ServiceClient>(ServiceClient.class)
	{
		@Override
		protected boolean annotationEquals(ServiceClient left, ServiceClient right)
		{
			return EqualsUtil.equals(left.value(), right.value());
		}
	};

	@Property(name = ServiceConfigurationConstants.NetworkClientMode, defaultValue = "false")
	protected boolean isNetworkClientMode;

	@Override
	protected ICascadedInterceptor handleServiceIntern(IBeanContextFactory beanContextFactory, IServiceContext beanContext,
			IBeanConfiguration beanConfiguration, Class<?> type, Set<Class<?>> requestedTypes)
	{
		Service serviceAnnotation = serviceAnnotationCache.getAnnotation(type);
		if (serviceAnnotation != null)
		{
			return handleServiceAnnotation(serviceAnnotation, beanContextFactory, beanContext, beanConfiguration, type);
		}
		AnnotationEntry<ServiceClient> serviceClientAnnotation = serviceClientAnnotationCache.getAnnotationEntry(type);
		if (serviceClientAnnotation != null)
		{
			return handleServiceClientAnnotation(serviceClientAnnotation, beanContextFactory, beanContext, beanConfiguration, type);
		}
		return super.handleServiceIntern(beanContextFactory, beanContext, beanConfiguration, type, requestedTypes);
	}

	protected String extractServiceName(IServiceContext beanContext, String serviceName, Class<?> type)
	{
		if (serviceName == null || serviceName.length() == 0)
		{
			serviceName = type.getSimpleName();
			if (serviceName.endsWith("Proxy"))
			{
				serviceName = serviceName.substring(0, serviceName.length() - 5);
			}
			if (serviceName.charAt(0) == 'I' && Character.isUpperCase(serviceName.charAt(1)))
			{
				serviceName = serviceName.substring(1);
			}
		}
		return serviceName;
	}

	protected ICascadedInterceptor handleServiceAnnotation(Service serviceAnnotation, IBeanContextFactory beanContextFactory, IServiceContext beanContext,
			IBeanConfiguration beanConfiguration, Class<?> type)
	{
		if (serviceAnnotation.customExport())
		{
			// Do nothing if the service wants to be exported by some special way anywhere else
			return null;
		}
		String beanName = beanConfiguration.getName();
		String serviceName = extractServiceName(beanContext, serviceAnnotation.name(), type);
		if (!isNetworkClientMode)
		{
			IMethodLevelBehavior<Annotation> behavior = createInterceptorModeBehavior(type);

			CacheInterceptor interceptor = new CacheInterceptor();
			if (beanContext.isRunning())
			{
				interceptor = beanContext.registerWithLifecycle(interceptor).propertyValue("ServiceName", serviceName).propertyValue("Behavior", behavior)
						.ignoreProperties("ProcessService").finish();
				beanContext.link(beanName).to(IServiceExtendable.class).with(serviceName);
			}
			else
			{
				beanContextFactory.registerWithLifecycle(interceptor).propertyValue("ServiceName", serviceName).propertyValue("Behavior", behavior)
						.ignoreProperties("ProcessService");
				beanContextFactory.link(beanName).to(IServiceExtendable.class).with(serviceName);
			}
			if (log.isInfoEnabled())
			{
				log.info("Registering application service '" + serviceName + "'");
			}
			return interceptor;
		}
		else
		{
			if (log.isInfoEnabled())
			{
				log.info("Registering application client stub '" + serviceName + "'");
			}
			if (beanContext.isRunning())
			{
				beanContext.link(beanName).to(IServiceExtendable.class).with(serviceName);
			}
			else
			{
				beanContextFactory.link(beanName).to(IServiceExtendable.class).with(serviceName);
			}
			return null;
		}
	}

	protected ICascadedInterceptor handleServiceClientAnnotation(AnnotationEntry<ServiceClient> serviceClientAnnotation,
			IBeanContextFactory beanContextFactory, IServiceContext beanContext, IBeanConfiguration beanConfiguration, Class<?> type)
	{
		String serviceName = extractServiceName(beanContext, serviceClientAnnotation.getAnnotation().value(), serviceClientAnnotation.getDeclaringType());

		IMethodLevelBehavior<Annotation> behavior = createInterceptorModeBehavior(type);

		CacheInterceptor interceptor = new CacheInterceptor();
		if (beanContext != null)
		{
			interceptor = beanContext.registerWithLifecycle(interceptor).propertyValue("ServiceName", serviceName).propertyValue("Behavior", behavior).finish();
			// beanContext.link(cacheInterceptorName).to(ICacheServiceByNameExtendable.class).with(serviceName);
		}
		else
		{
			beanContextFactory.registerWithLifecycle(interceptor).propertyValue("ServiceName", serviceName).propertyValue("Behavior", behavior);
			// beanContextFactory.link(cacheInterceptorName).to(ICacheServiceByNameExtendable.class).with(serviceName);
		}

		if (log.isInfoEnabled())
		{
			log.info("Creating application service stub for service '" + serviceName + "' accessing with '"
					+ serviceClientAnnotation.getDeclaringType().getName() + "'");
		}
		return interceptor;
	}

	protected String buildCacheInterceptorName(String serviceName)
	{
		return "cacheInterceptor." + serviceName;
	}

	@Override
	protected Annotation lookForAnnotation(AnnotatedElement member)
	{
		Annotation annotation = super.lookForAnnotation(member);
		if (annotation != null)
		{
			return annotation;
		}
		return member.getAnnotation(Cached.class);
	}
}
